### PR TITLE
DSEGOG-51 DSEGOG-55 DSEGOG-60 DSEGOG-63 DSEGOG-66 Initial plotting

### PR DESCRIPTION
**Pull request by louise-davies**
## Description
This either partially tackles or completely implements all the issues mentioned in the title. This uses `react-chartjs-2` and thus `chart.js` as the underlying chart library as when I examined it it seemed to support all the features we needed. For this initial PR, I have added the chart itself, the config menu (this is a WIP based on the balsamiq wireframes - most notably the `hours`, `points`, `min` and `max` boxes control nothing right now) and a "window" that holds both the plot and the settings menu in a collapsable drawer.

I have just made a basic `ConnectedPlot` component to pass the timestamp and shot numbers from the test data to the plot, but this area of the code will likely change in the future as there's all the channel selection stuff so don't examine that too closely!

I haven't added any e2e tests yet as the only real thing that needs e2e testing is the graph itself (as we can't test this in jest). We can use image snapshot testing using a library like https://github.com/jaredpalmer/cypress-image-snapshot to test the graphs look like we expect, but first we need to be able to pass stable test data to cypress as otherwise it will always fail 😅 

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] Play about with the plotting tab & settings controls

## Agile board tracking
Closes DSEGOG-60, DSEGOG-66, DSEGOG-55
Partially addresses DSEGOG-63 (logic for selecting X axis scale when it's not time is not there yet as this PR just assumes X is time),  DSEGOG-51 (we can't select different channels for the Y axes yet - I figured adding channel selection for both X & Y axes made sense in a separate PR), tiny bit of DSEGOG-67 (configurable title)

### CODE REVIEW

File: src/api/records.tsx [hunk 1]
```diff
@@ -80,6 +85,34 @@ export const useRecordsPaginated = (): UseQueryResult<
   );
 };
 
+// TODO: should we integrate this with useRecordsPaginated? and have options to
+// pass the recordRow select function and which of the query params to apply?
```
Comment (sam-glendenning) [inline src/api/records.tsx:89]: So have approprate `useQueryResult` options for fetching pages of records (for use in Data tab) and options for fetching records by date range (for use in Plot tab)? Yeah I'm down for that. We already do it in `channels.tsx` and I quite like how it looks. Feel free to spread out the code into separate files if it makes sense to do that. For example, we can keep `records.tsx` purely for fetching data and other functions in other files in `src/api` for curating the options and calls to the `useRecords` function.

File: package.json [hunk 2]
```diff
@@ -18,13 +18,16 @@
...
+    "chart.js": "^3.8.2",
+    "chartjs-adapter-date-fns": "^2.0.0",
     "date-fns": "2.29.1",
     "history": "5.3.0",
     "i18next": "21.8.14",
     "immutability-helper": "3.1.1",
     "loglevel": "1.8.0",
     "react": "18.2.0",
     "react-beautiful-dnd": "13.1.0",
+    "react-chartjs-2": "^4.3.1",
```
Comment (sam-glendenning) [inline package.json]: Should these dependencies be pinned?

File: src/plotting/plot.component.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      aria-label={`${title} plot`}
+    />
+  );
+};
+
+interface PlotProps {
+  title: string;
+  type: PlotType;
+  XAxisSettings: AxisSettings;
+  YAxesSettings: AxisSettings;
```
Comment (sam-glendenning) [inline src/plotting/plot.component.tsx:76]: Minor typo, change to `YAxisSettings`

File: package.json (hunk 2 above)
Comment (sam-glendenning) [reply package.json]: Also noticed there's a new minor version of `chart.js`, 3.9.1. Worth updating to give it a try

File: src/plotting/plot.component.tsx (hunk 3 above)
Comment (louise-davies) [reply src/plotting/plot.component.tsx:76]: This isn't a typo ;) There's potentially two Y axes (it's in a later requirement), hence using the plural axes

File: src/api/records.tsx (hunk 1 above)
Comment (louise-davies) [reply src/api/records.tsx:89]: I probably will separate them then - it also makes it easier to mock just the data fetching functions whilst leaving the react-query functionality if they're in separate files (see `dg-download`).

File: src/plotting/plot.component.tsx (hunk 3 above)
Comment (sam-glendenning) [reply src/plotting/plot.component.tsx:76]: Oh yes, you're right, my bad. I assume they'd have the same settings applied to both? My mind is still trying to get around two y-axes

Comment (louise-davies) [reply src/plotting/plot.component.tsx:76]: The two axes can have independent settings - not sure how they want this to work in the UI but it's definitely possible using chart.js: https://react-chartjs-2.js.org/examples/multiaxis-line-chart/

File: src/plotting/plotWindow.component.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  const [plotTitle, setPlotTitle] = React.useState('');
+  const [plotType, setPlotType] = React.useState<PlotType>('scatter');
+  const [XAxisSettings, setXAxisSettings] = React.useState<AxisSettings>({
+    scale: 'time',
+  });
+  const [YAxesSettings, setYAxesSettings] = React.useState<AxisSettings>({
+    scale: 'linear',
+  });
+
+  const [open, setOpen] = React.useState(false);
```
Comment (sam-glendenning) [inline src/plotting/plotWindow.component.tsx]: There's advantages to both but is it preferable to have the drawer open by default? This mimics the behaviour of eCat and allows the user to easily see what they can do without knowing to find and click the small settings icon in the top left

**Comment by sam-glendenning** [comment]
Might just be me but I found I always needed to scroll down to see the x-axis in full-screen view of the plot. Even when zooming out from 100%. Could this be changed?

**Comment by louise-davies** [comment]
> Might just be me but I found I always needed to scroll down to see the x-axis in full-screen view of the plot. Even when zooming out from 100%. Could this be changed?

Plots will eventually be in new windows anyway - so I deliberately didn't put too much styling/sizing stuff into this PR as it's just a temporary measure having it in the main window

File: src/plotting/plotWindow.component.tsx (hunk 4 above)
Comment (louise-davies) [reply src/plotting/plotWindow.component.tsx]: true - I'll flip this around

**Comment by sam-glendenning** [comment]
Just out of curiosity, how are the data points in this demo in the line graph connected? I'm completely confused as to how, for example, shotnum 527 on Jan 5 2024 is connected to shotnum 921, Jan 19 2024

**Comment by sam-glendenning** [comment]
I might argue that DSEGOG -55 won't be closed in its entirety as some parts of the plotSettings have yet to be implemented. I'm also aware of what Jira will automatically close when this PR is merged since it mentions so many issues and that's enough to move them to done when PR is merged (so either manually move them back on the sprint board or change their reference to include a space in their name as I _think_ that also does the trick)

Alternatively, we can make new issues for the things remaining, such as hooking up the min and max fields

**Comment by louise-davies** [comment]
> Just out of curiosity, how are the data points in this demo in the line graph connected? I'm completely confused as to how, for example, shotnum 527 on Jan 5 2024 is connected to shotnum 921, Jan 19 2024

I presume they're connected in the order they're passed to the chart in - in practice this means we'll need to specify that we want to order the results back from the API in `X axis` order ascending

File: src/api/records.tsx (hunk 1 above)
Comment (louise-davies) [reply src/api/records.tsx:89]: @sam-glendenning I wonder if it's worth leaving this TODO until after the data fetching PR is in? So this PR can go in without waiting for it

Comment (sam-glendenning) [reply src/api/records.tsx:89]: I'm happy with that. Make a Jira issue if necessary so we don't lose track :)

**Review by sam-glendenning — APPROVED** [review]
Yep, looks good. I'm happy for DSEGOG-60 and DSEGOG-66 to be totally closed by this PR but I'd say other issues still need some work done on them. Some of these issues (notably DSEGOG-51 and DSEGOG-63) are being closed by my [WIP PR here](https://github.com/ral-facilities/operationsgateway/pull/62). I'm unsure when DSEGOG-55 can be called "done" because that can encompass editing _any_ specifications of the plot or _all_ specifications.

You know what, I've changed my mind. DSEGOG-55 is also closed by this PR in my opinion. Other issues can pick up the rest.

**Comment by louise-davies** [comment]
@sam-glendenning I agree - I'll sort out the JIRA to ensure the correct issues are marked as done vs in progress & add an issue to the board to refactor `useRecords` with plots once your PR gets merged

**Pull request merged**